iac-extensions/docker/src/main/java/org/sonar/iac/docker/symbols/Symbol.java

```diff
@@ -56,4 +56,16 @@ public List<Usage> usages() {
     return usages;
   }
 
+  public Scope.Kind lastDeclarationScope() {
+    Scope.Kind declarationScope = Scope.Kind.IMAGE;
+    for (int i = usages.size() - 1; i >= 0; i--) {
+      Usage usage = usages.get(i);
+      if (usage.kind().equals(Usage.Kind.ASSIGNMENT)) {
+        declarationScope = usage.scope().kind();
+        // we stop at the first assignment we find, as we store usages in the same order they appear in Dockerfile
+        break;
+      }
+    }
+    return declarationScope;
+  }
```

Comment:
I'm wondering what result do you get for a variable that is used but never declared? Do we have a symbol for it in that case? Do this methis return an `IMAGE` scope in that case? Shouldn't we get null instead? (+add @CheckForNull)

Reply:
We don't create symbols for such variables, AFAIU. In `DockerSymbolVisitor` we only add symbols when visiting assignment nodes, that's why we can assume that if there is no symbol for a variable, then it's not a Dockerfile-level variable, but probably a shell env.